pkg/envoy/grpc.go

```diff
@@ -0,0 +1,596 @@
+					}}}},
+				}},
+				DeprecatedV1: &envoy_api.Filter_DeprecatedV1{
+					Type: "read",
+				},
+			}},
+		}},
+		ListenerFilterChain: []*envoy_api.Filter{{
+			Name: "bpf_metadata",
+			Config: &structpb.Struct{Fields: map[string]*structpb.Value{
+				"deprecated_v1": {&structpb.Value_BoolValue{BoolValue: true}},
+				"value": {&structpb.Value_StructValue{StructValue: &structpb.Struct{Fields: map[string]*structpb.Value{
+					"is_ingress": {&structpb.Value_BoolValue{BoolValue: false}},
+				}}}},
+			}},
+			DeprecatedV1: &envoy_api.Filter_DeprecatedV1{
+				Type: "accept",
+			},
+		}},
+	}
+
+	ldsServer.listeners = make(map[string]*Listener)
+	ldsServer.envoyResources = make(map[string]*Listener)
+
+	envoy_api.RegisterListenerDiscoveryServiceServer(ldsServer.glds, ldsServer)
+	// Register reflection service on gRPC server.
+	reflection.Register(ldsServer.glds)
+
+	return ldsServer
+}
+
+func (s *LDSServer) addListener(name string, port uint16, l7rules policy.L7DataMap, isIngress bool) {
+	s.listenersMutex.Lock()
+	log.Debug("Envoy: addListener ", name)
+
+	// Bail out if this listener already exists
+	if _, ok := s.listeners[name]; ok {
+		log.Fatalf("Envoy: addListener: Listener %s already exists!", name)
+	}
+
+	resourceName := "RDS_" + name + "_" + strconv.FormatUint(s.currentVersion, 10)
+
+	listener := &Listener{
+		proxyPort:     port,
+		l7rules:       l7rules,
+		listenerConf:  proto.Clone(s.listenerProto).(*envoy_api.Listener),
+		StreamControl: makeStreamControl(resourceName),
+	}
+
+	// Fill in the listener-specific parts
+	listener.listenerConf.Name = name
+	listener.listenerConf.Address.GetSocketAddress().PortSpecifier = &envoy_api.SocketAddress_PortValue{PortValue: uint32(port)}
+	listener.listenerConf.FilterChains[0].Filters[0].Config.Fields["rds"].GetStructValue().Fields["route_config_name"] = &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: resourceName}}
+	if isIngress {
+		listener.listenerConf.ListenerFilterChain[0].Config.Fields["value"].GetStructValue().Fields["is_ingress"].GetKind().(*structpb.Value_BoolValue).BoolValue = true
+	}
+	s.listeners[name] = listener
+	s.envoyResources[resourceName] = listener
+	s.listenersMutex.Unlock()
+	s.bumpVersion()
+}
+
+func (s *LDSServer) updateListener(name string, l7rules policy.L7DataMap) {
+	s.listenersMutex.Lock()
+	log.Debug("Envoy: updateListener ", name)
+	defer s.listenersMutex.Unlock()
+	l := s.listeners[name]
+	// Bail out if this listener does not exist
+	if l == nil {
+		log.Fatalf("Envoy: updateListener: Listener %s does not exist", name)
+	}
+
+	l.bumpVersionFunc(func() { l.l7rules = l7rules })
+}
+
+func (s *LDSServer) removeListener(name string) {
+	s.listenersMutex.Lock()
+	log.Debug("Envoy: removeListener ", name)
+	l := s.listeners[name]
+	// Bail out if this listener does not exist
+	if l == nil {
+		log.Fatalf("Envoy: removeListener: Listener %s does not exist", name)
```

Comment:
Updating or removing a redirect that does not exist would be an internal logic violation in cilium-agent and should be noticed rather than buried in logs.

If these can fail then they should also return an error, and the callers should be changed to deal with the failures. Currently these are called when bpf is being regenerated (part of the writeHeaderfile()) and we would rather not fail at that point.